text/0000-socket-timeouts.md

```diff
@@ -0,0 +1,109 @@
+- Feature Name: socket_timeouts
+- Start Date: 2015-04-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add sockopt-style timeouts to `std::net` types.
+
+# Motivation
+
+Currently, operations on various socket types in `std::net` block
+indefinitely (i.e., until the connection is closed or data is
+transferred). But there are many contexts in which timing out a
+blocking call is important.
+
+The [goal of the current IO system][io-reform] is to gradually expose
+cross-platform, blocking APIs for IO, especially APIs that directly
+correspond to the underlying system APIs. Sockets are widely available
+with nearly identical system APIs across the platforms Rust targets,
+and this includes support for timeouts via [sockopts][sockopt].
+
+So timeouts are well-motivated and well-suited to `std::net`.
+
+# Detailed design
+
+The proposal is to *directly expose* the timeout functionality
+provided by [`setsockopt`][sockopt], in much the same way we currently
+expose functionality like `set_nodelay`:
+
+```rust
+impl TcpStream {
+    pub fn set_read_timeout(&self, dur: Duration) -> io::Result<()> { ... }
+    pub fn set_write_timeout(&self, dur: Duration) -> io::Result<()> { ... }
+}
+
+impl UdpSocket {
+    pub fn set_read_timeout(&self, dur: Duration) -> io::Result<()> { ... }
+    pub fn set_write_timeout(&self, dur: Duration) -> io::Result<()> { ... }
+}
+```
+
+These methods take an amount of time in the form of a `Duration`,
+which is [undergoing stabilization][duration-reform]. They are
+implemented via straightforward calls to `setsockopt`.
```

Comment:
In the interest of the "detailed" part of the "Detailed design", could this also mention that the specific options being used are `SO_SNDTIMEO` and `SO_RCVTIMEO`?